go/vt/vtgate/healthcheck.go

```diff
@@ -0,0 +1,279 @@
+package vtgate
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/golang/glog"
+	pbq "github.com/youtube/vitess/go/vt/proto/query"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/tabletserver/tabletconn"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// AddrUpdate updates about an endpoint and whether it is up.
+type AddrUpdate struct {
+	EndPoint *pbt.EndPoint
+	IsUp     bool
+}
+
+// NewHealthCheck creates a new HealthCheck object.
+func NewHealthCheck(endPoints []*pbt.EndPoint, addrUpdate <-chan *AddrUpdate, connTimeout time.Duration, retryDelay time.Duration) *HealthCheck {
+	hc := &HealthCheck{
+		addrToConns: make(map[string]*healthCheckConn),
+		targetToEPs: make(map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint),
+		connTimeout: connTimeout,
+		retryDelay:  retryDelay,
+	}
+	hc.initConns(endPoints)
+	go hc.onAddrUpdate(addrUpdate)
+	return hc
+}
+
+// HealthCheck performs health checking and notifies downstream components about any changes.
+type HealthCheck struct {
+	mu          sync.RWMutex
+	addrToConns map[string]*healthCheckConn                              // addrToConns maps from address to the healthCheckConn object.
+	targetToEPs map[string]map[string]map[pbt.TabletType][]*pbt.EndPoint // targetToEPs maps from keyspace/shard/tablettype to a list of endpoints.
+	connTimeout time.Duration
+	retryDelay  time.Duration
+}
+
+// healthCheckConn contains details about an endpoint.
+type healthCheckConn struct {
+	mu         sync.RWMutex
+	target     *pbq.Target
+	cancelFunc context.CancelFunc
+	conn       tabletconn.TabletConn
+	stats      *pbq.RealtimeStats
+}
+
+// initConns starts health checking for the initial set of endpoints.
+// It waits till all endpoints are tried to connect once.
```

Comment:
While this would work, I think it is too big of a hammer, we only need 'a few' instances to be able to serve, not all of them, and I don't think the right criteria is 'we tried everybody once'.

I think the real thing we're waiting for at startup is probably more complicated, and should be a different module that makes the checks and waits until this is all ready. This would make this module simpler, and the policy in a different place and easier to change.